StrongInject.Generator/Operation.cs

```diff
@@ -2,23 +2,12 @@
 
 namespace StrongInject.Generator
 {
-    internal class Operation
-    {
-        public Operation(Statement statement, Disposal? disposal, List<Operation> dependencies, bool canDisposeLocally, AwaitStatement? awaitStatement = null)
-        {
-            Statement = statement;
-            Disposal = disposal;
-            Dependencies = dependencies;
-            AwaitStatement = awaitStatement;
-            CanDisposeLocally = canDisposeLocally;
-        }
-
-        public Statement Statement { get; }
-        public Disposal? Disposal { get; }
-        public AwaitStatement? AwaitStatement { get; }
-        public bool CanDisposeLocally { get; }
-        public List<Operation> Dependencies { get; }
-    }
+    internal record Operation(
```

Comment:
It's possible the reason I didn't make this a record is because it uses reference equality - probably in the stage in DisposalLowerer where we order operations. Do you mind checking to make sure?

Reply:
All GeneratorTests pass after the change. I found only this so far. `==` and `!=` operators and `Equals` method are not directly used on `Operation` instances. Other than that, I'm not 100% sure but I think these are the only reference equality comparisons on `Operation` instances:

https://github.com/YairHalberstadt/stronginject/blob/2143d9f27dfdf7269b4de3e60dd198c4374cd586/StrongInject.Generator/Visitors/LoweringVisitor.cs#L282

https://github.com/YairHalberstadt/stronginject/blob/2143d9f27dfdf7269b4de3e60dd198c4374cd586/StrongInject.Generator/Visitors/LoweringVisitor.cs#L348

If you like, I can refactor back to a class and write a manual `Operation WithDisposal(Disposal disposal)` method.

Reply:
Perhaps you can override equality to explicitly use ReferenceEquals?

Reply:
Ah, yes, that is possible:

```cs
internal sealed record Operation(
    Statement Statement,
    Disposal? Disposal,
    List<Operation> Dependencies,
    bool CanDisposeLocally,
    AwaitStatement? AwaitStatement = null) : IEquatable<Operation>
{
    public bool Equals(Operation other) => ReferenceEquals(this, other);

#pragma warning disable RS1024 // "Compare symbols correctly" is buggy
    public override int GetHashCode() => RuntimeHelpers.GetHashCode(this);
#pragma warning restore RS1024
}
```

If this was .NET 5, I'd recommend https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.referenceequalitycomparer.instance to the hash set instead since I see this as a usage concern rather than a definitional concern, so it keeps the coupling from being so spread out.

I wish I had proposed that API a long time ago. :)

Reply:
I can't resolve/unresolve, but I'm keeping this as a TODO for myself since I think I might end up not using `with` expressions in the local function formulation of our plan.